db_stress_tool/expected_state.cc

```diff
@@ -196,23 +252,325 @@ Status FileExpectedStateManager::Open() {
+  Status s =
+      CopyFile(FileSystem::Default(), latest_file_path, state_file_temp_path,
+               0 /* size */, false /* use_fsync */);
+  if (s.ok()) {
+    s = FileSystem::Default()->RenameFile(state_file_temp_path, state_file_path,
+                                          IOOptions(), nullptr /* dbg */);
+  }
+  SequenceNumber old_saved_seqno;
+  if (s.ok()) {
+    old_saved_seqno = saved_seqno_;
+    saved_seqno_ = seqno;
+  }
+
+  // If there is a crash now, i.e., after "<seqno>.state" was created but before
+  // "<seqno>.trace" is created, it will be treated as if "<seqno>.trace" were
+  // present but empty.
+
+  // Create "<seqno>.trace" directly. It is initially empty so no need for
+  // tempfile.
+  std::unique_ptr<TraceWriter> trace_writer;
+  if (s.ok()) {
+    EnvOptions soptions;
+    // Disable buffering so traces will not get stuck in application buffer.
+    soptions.writable_file_max_buffer_size = 0;
+    s = NewFileTraceWriter(Env::Default(), soptions, trace_file_path,
+                           &trace_writer);
+  }
+  if (s.ok()) {
+    s = db->StartTrace(TraceOptions(), std::move(trace_writer));
+  }
+
+  // Delete old state/trace files. Deletion order does not matter since we only
+  // delete after successfully saving new files, so old files will never be used
+  // again, even if we crash.
+  if (s.ok()) {
+    if (old_saved_seqno != kMaxSequenceNumber &&
+        old_saved_seqno != saved_seqno_) {
+      s = Env::Default()->DeleteFile(
+          GetPathForFilename(ToString(old_saved_seqno) + kStateFilenameSuffix));
+    }
+  }
+  if (s.ok()) {
+    if (old_saved_seqno != kMaxSequenceNumber &&
+        old_saved_seqno != saved_seqno_) {
+      s = Env::Default()->DeleteFile(
+          GetPathForFilename(ToString(old_saved_seqno) + kTraceFilenameSuffix));
+    }
+  }
+  return s;
+}
+#else   // ROCKSDB_LITE
+Status FileExpectedStateManager::SaveAtAndAfter(DB* /* db */) {
+  return Status::NotSupported();
+}
+#endif  // ROCKSDB_LITE
+
+bool FileExpectedStateManager::HasHistory() {
+  return saved_seqno_ != kMaxSequenceNumber;
+}
+
+#ifndef ROCKSDB_LITE
+Status FileExpectedStateManager::Restore(DB* db) {
+  // An `ExpectedStateTraceRecordHandler` applies a configurable number of
+  // write operation trace records to the configured expected state.
+  class ExpectedStateTraceRecordHandler : public TraceRecord::Handler,
+                                          public WriteBatch::Handler {
+   public:
+    ExpectedStateTraceRecordHandler(uint64_t max_write_ops,
+                                    ExpectedState* state)
+        : max_write_ops_(max_write_ops), state_(state) {}
+
+    ~ExpectedStateTraceRecordHandler() {
+      assert(num_write_ops_ == max_write_ops_);
+    }
+
+    Status Handle(const WriteQueryTraceRecord& record,
+                  std::unique_ptr<TraceRecordResult>* /* result */) override {
+      if (num_write_ops_ == max_write_ops_) {
+        return Status::OK();
+      }
```

Comment:
> I'm not understanding how it's able to sync up the replayed trace and the DB state. Can you explain? I don't see for example a seqno filter or threshold on what we apply to expected state.

This returns without applying the write op once we've reached `max_write_ops_`. The client configures `max_write_ops_` by subtracting the base sequence number of the trace determined by the filename ("X.trace" has base seqno X) from the post-recovery `GetLatestSequenceNumber()`.

Reply:
OK thanks. Of course it looks less hidden now that you've showed it to me, but I looked for it for a while.